graph/src/firehose/endpoints.rs

```diff
@@ -1,826 +1,960 @@
+            SubgraphLimit::Disabled | SubgraphLimit::Limit(0) => AvailableCapacity::Unavailable,
+            SubgraphLimit::Limit(total) => {
+                let total = *total;
+                if current >= total {
+                    return AvailableCapacity::Unavailable;
+                }
+
+                let used_percent = current * 100 / total;
+
+                // If total is low it can vary very quickly so we can consider 50% as the low threshold
+                // to make selection more reliable
+                let threshold_percent = if total <= LOW_VALUE_THRESHOLD {
+                    LOW_VALUE_USED_PERCENTAGE
+                } else {
+                    HIGH_VALUE_USED_PERCENTAGE
+                };
+
+                if used_percent < threshold_percent {
+                    return AvailableCapacity::High;
+                }
+
+                AvailableCapacity::Low
+            }
+            _ => AvailableCapacity::High,
+        }
+    }
+
+    pub fn has_capacity(&self, current: usize) -> bool {
+        match self {
+            SubgraphLimit::Unlimited => true,
+            SubgraphLimit::Limit(limit) => limit > &current,
+            SubgraphLimit::Disabled => false,
+        }
+    }
+}
+
+impl Display for FirehoseEndpoint {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Display::fmt(self.provider.as_str(), f)
+    }
+}
+
+impl FirehoseEndpoint {
+    pub fn new<S: AsRef<str>>(
+        provider: S,
+        url: S,
+        token: Option<String>,
+        key: Option<String>,
+        filters_enabled: bool,
+        compression_enabled: bool,
+        subgraph_limit: SubgraphLimit,
+        endpoint_metrics: Arc<EndpointMetrics>,
+        genesis_decoder: Box<dyn GenesisDecoder>,
+    ) -> Self {
+        let uri = url
+            .as_ref()
+            .parse::<Uri>()
+            .expect("the url should have been validated by now, so it is a valid Uri");
+
+        let endpoint_builder = match uri.scheme().unwrap_or(&Scheme::HTTP).as_str() {
+            "http" => Channel::builder(uri),
+            "https" => Channel::builder(uri)
+                .tls_config(ClientTlsConfig::new())
+                .expect("TLS config on this host is invalid"),
+            _ => panic!("invalid uri scheme for firehose endpoint"),
+        };
+
+        // These tokens come from the config so they have to be ascii.
+        let token: Option<MetadataValue<Ascii>> = token
+            .map_or(Ok(None), |token| {
+                let bearer_token = format!("bearer {}", token);
+                bearer_token.parse::<MetadataValue<Ascii>>().map(Some)
+            })
+            .expect("Firehose token is invalid");
+
+        let key: Option<MetadataValue<Ascii>> = key
+            .map_or(Ok(None), |key| {
+                key.parse::<MetadataValue<Ascii>>().map(Some)
+            })
+            .expect("Firehose key is invalid");

```

Comment:
All these expect and panics will cause a configuration error to stop `graph-node` from starting, and AFAICT, we don't even have tooling to check that part of the configuration. So a typo like `url = "htps://some.where.com/"` will make it impossible to start `graph-node`. These issues should result in an error, and that part of the config should be ignored, but the rest of `graph-node` should start up successfully.

Reply:
same as the previous comment, this wasn't changed, this code has been reviewed elsewhere and hasn't been modified

Reply:
I didn't realize that this code was moved over. I went through the PR to look at places where it can produce a panic, as we've had several instances now where panics can cause huge operational issues and we need to avoid them as much as possible.